Fixed #1166 - Currency Format

sunarcrajneesh opened the pull request:
## Issue Reference
<!--- Please mention issue #id or use a comma if your pull request solves multiple issues. -->
Fixes #1166 -  Currency format Enhancement
## Description
<!--- Please describe your changes in detail. -->

- We made changes to the database currencies table by altering it and adding new columns for decimal separator, thousand separator, and currency position. These changes are reflected in the edit and add forms, which can be accessed at /admin/settings/currencies.
- We updated the models table to make these new fields fillable.
- We also made changes in the CurrencyController, where we created a currencyPosition array to populate dropdown options in the form. and in store and update methods this allows us to add and edit fields seamlessly.
- To improve the formatting of currency values, we introduced a new function called 'formatPriceWithSymbol' in the core.php file. This function helps in formatting currency and adjusting the position of the currency symbol. Furthermore, we wrapped the currency symbol in an HTML tag with a class, which was created for future styling customizations.
-  For styling purposes, we created a new class in the public/css/currency-symbol-style.css file and linked this file in packages/Webkul/Shop/src/Resources/views/components/layouts/index.blade.php to apply the styles in the user panel.
- To enhance the formatting of the total price of bundle products, we implemented a new API in packages/Webkul/Shop/src/Routes/api.php and utilized it in packages/Webkul/Shop/src/Resources/views/products/view/types/bundle.blade.php within the v-product-bundle-options component.
- In packages/Webkul/Shop/src/Resources/views/products/prices/bundle.blade.php, we made adjustments to the HTML structure to avoid using CSS classes on both the currency symbol and price separately.
- Lastly, to render HTML tags attached to the currency symbol, we using {!! !!} or v-html in Blade files instead of {{ }} or v-text."

## How To Test This?
<!--- Please describe in detail how to test the changes made in this pull request. -->

- This pull request introduces a feature that enhances currency formatting. It now supports customizing both thousand separators and decimal separators, which can be configured via the settings form. In cases where these fields are left empty, a space separator is automatically used. Additionally, this feature allows for the display of the currency symbol in the selected position.
- The styling of the currency symbol is managed through CSS, which has been added to the 'currency-symbol-style' class in the 'public/css/currency-symbol-style.css' file."

## Documentation
- [ ] My pull request requires an update on the documentation repository.
<!--- Please describe in detail what needs to be changed. --->

Amitk-webkul commented:
@sunarcrajneesh Getting a few issues. Please check

A. Not able to create/edit currency, has errors in the console. [Here](https://prnt.sc/dvALtDgjgA3g)
B. Getting Spelling mistake in Decimal Separator. [Here](https://prnt.sc/Yu2KJvzPpN2V)

sunarcrajneesh commented:
Able to create/edit currency without any errors.
Please let us know the version you are using.

Amitk-webkul commented:
@sunarcrajneesh With the master branch.

sunarcrajneesh commented:
We have verified the situation, and we have not encountered any errors so far. Would you kindly confirm whether you have executed the migrate command?
![Screenshot from 2023-10-04 13-08-27](https://github.com/bagisto/bagisto/assets/45708520/e55aa4f8-7f8b-4b9a-a349-7951d224bfca)

sunarcrajneesh commented:
We have resolved the minicart exception, fixed that typo, and also translation issues related to 'starting price' in grouped products and 'As low as' in configurable products.

devansh-webkul commented on packages/Webkul/Core/src/Core.php:
```diff
@@ -252,7 +252,7 @@ public function getRequestedChannelCode($fallback = true)
         $channelCode = request()->get('channel');
 
-        if (! $fallback) {
+        if (!$fallback) {
```
Same here.

sunarcrajneesh replied:
Done.

devansh-webkul commented on packages/Webkul/Core/src/Core.php:
```diff
@@ -605,21 +605,81 @@ public function formatPrice($price, $currencyCode = null)
-            }
 
+                return $this->formatPriceWithSymbol($price, $currency, $currency->symbol, $formatter);
+                // _________________________________________
+
+                // return $formatter->formatCurrency($price, $currency->code);
```
Remove unnecessary comments.

sunarcrajneesh replied:
Done.

devansh-webkul commented on packages/Webkul/Core/src/Core.php:
```diff
@@ -605,21 +605,81 @@ public function formatPrice($price, $currencyCode = null)
+    }
+
+    /**
+     *
```
Doc comment should start from here, no extra line should be in the start. And after one info there should be one more extra line.

sunarcrajneesh replied:
Done.

devansh-webkul commented on packages/Webkul/Core/src/Core.php:
```diff
@@ -605,21 +605,81 @@ public function formatPrice($price, $currencyCode = null)
+     *
+     * customize the thousand separator and decimal separator.
+     * Additionally, repositioning of the currency symbol and the inclusion of an HTML tag for further styling.
+     * @param  float  $price
+     * @param Webkul\Core\Models\Currency $currency
+     * @param string $symbol
+     * @param \NumberFormatter $formatter 
```
https://laravel.com/docs/10.x/contributions#phpdoc

sunarcrajneesh replied:
Thanks for sharing the link! We have followed it.

devansh-webkul commented on packages/Webkul/Shop/src/Routes/api.php:
```diff
@@ -103,4 +104,7 @@
         });
     });
+
+    // for formatting total price in bundle product
+    Route::get('/formatPrice/{slug1}/{slug2}/', [Core::class, 'formatPrice'])->name('shop.api.format_price');
```
Don't add the core class directly. Use controller.

sunarcrajneesh replied:
Done.

devansh-webkul commented on packages/Webkul/Admin/src/Http/Controllers/Settings/CurrencyController.php:
```diff
@@ -28,8 +28,11 @@ public function index()
         }
-
-        return view('admin::settings.currencies.index');
+        $currencyPosition = [
+            'left' => 'Left', 'left_with_space' => 'Left With Space',
+            'right' => 'Right', 'right_with_space' => 'Right With Space',
+        ];
```
- All key values, should be in individual lines.

devansh-webkul replied:
Also, add translations.

sunarcrajneesh replied:
Both tasks have been completed. All key values are now on individual lines, and translations have been added.

devansh-webkul commented on packages/Webkul/Admin/src/Http/Controllers/Settings/CurrencyController.php:
```diff
@@ -28,8 +28,11 @@ public function index()
         if (request()->ajax()) {
             return app(CurrencyDataGrid::class)->toJson();
         }
```
One line after closing tag.

sunarcrajneesh replied:
Done.

devansh-webkul commented on packages/Webkul/Admin/src/Http/Controllers/Settings/CurrencyController.php:
```diff
@@ -28,8 +28,11 @@ public function index()
+            'left' => 'Left', 'left_with_space' => 'Left With Space',
+            'right' => 'Right', 'right_with_space' => 'Right With Space',
+        ];
+        return view('admin::settings.currencies.index', compact('currencyPosition'));
```
One line before the return statement.

sunarcrajneesh replied:
Done.

devansh-webkul commented on public/css/currency-symbol-style.css:
```diff
@@ -0,0 +1,6 @@
+/* currency-symbol-style.css
+/* This style is applied on currency symbol */
+.currency-symbol-style {
+    color: green;
+    /* border: solid red 1px; */
+    } 
```
Why do we need this?

sunarcrajneesh replied:
We provided this for styling on the currency symbol.

devansh-webkul replied:
Remove this also, we are not giving any styling because this will force us to change the directive to v-html.

sunarcrajneesh replied:
Styling has been removed, and the files where 'v-html' was used have been restored to their previous state.

devansh-webkul commented on packages/Webkul/Shop/src/Resources/views/checkout/onepage/summary.blade.php:
```diff
@@ -29,21 +29,20 @@ class="flex gap-x-[15px] pb-[20px]"
 
                             <p class="mt-[10px] text-[18px] font-medium max-sm:text-[14px] max-sm:font-normal">
-                                @{{ item.formatted_price }} X @{{ item.quantity }}
+                                <span v-html="item.formatted_price"> </span> X @{{ item.quantity }}
```
Also remove the `v-html` fro everywhere, we are not giving formating at this stage.

devansh-webkul replied:
Just add left and right switch codes, don't add any HTML that needs to change the v-html.

devansh-webkul replied:
Yeah, i know, so we are not giving any styling at this stage. Just remove the last pointer which doing formating and stylings.

sunarcrajneesh replied:
Styling has been removed, and the files where 'v-html' and {!! !!} were used have been restored to their previous state.

Amitk-webkul commented:
@sunarcrajneesh Getting a few issues please check.

1. On the Add coupon button HTML code appears. [Here](https://prnt.sc/WT5FtQFN2Nde)
2. In other Languages currency signs appear twice. [Here](https://prnt.sc/UZo7oZbfIVq8)
3. If the currency symbol is removed/deleted from the admin side, On the Shop it shows both the currency symbol and code. [Here](https://prnt.sc/7hMg6oVj_lJb)

Amitk-webkul commented:
@sunarcrajneesh, Is there any update?

sunarcrajneesh commented:
@Amitk-webkul,
We have addressed the styling point related to the currency symbol and implemented the required file changes, following the suggestions made by @devansh-webkul.
After removing the currency symbol, it is no longer displaying in the shop, and we have also added the translations.

Amitk-webkul commented:
@sunarcrajneesh, Okay Thank you for confirming.

Amitk-webkul commented:
@sunarcrajneesh, Is there any update?

sunarcrajneesh commented:
@Amitk-webkul,
As per our earlier communication, we'd like to inform you that the reported bug has been successfully resolved on our end, and this PR also resolves #8605.

Amitk-webkul commented:
@sunarcrajneesh Please resolve the conflicts.

sunarcrajneesh commented:
@Amitk-webkul Conflicts resolved.

Amitk-webkul commented:
@sunarcrajneesh,  Please resolve the conflicts

Please rename the visit table as we have already used the visit table. [Here](https://prnt.sc/3JINSe9h4pCC)

sunarcrajneesh commented:
@Amitk-webkul,
Conflicts resolved. Please verify once more as we are not using the 'visit' table.

Amitk-webkul commented:
@sunarcrajneesh, It is working fine. Please add a fix comment to the review comments.

sunarcrajneesh commented on packages/Webkul/Core/src/Core.php:
```diff
@@ -161,7 +161,7 @@ public function getCurrentChannel()
         ])->first();
 
-        if (! $this->currentChannel) {
+        if (!$this->currentChannel) {
```
Applied with --dirty flag. Thanks for the guidance.

devansh-webkul commented:
Hi @sunarcrajneesh,

Can you please update the branch? As commits are too much behind.

sunarcrajneesh commented:
@devansh-webkul,
Sure, we'll update the branch to sync with the latest commits.

sunarcrajneesh commented:
@devansh-webkul,
We have updated the branch, please check it.

devansh-webkul commented:
@sunarcrajneesh,

Pint test is failing. Please run pint command on your project.

devansh-webkul commented:
I will take some more time on this PR as it relates to currency and I need to check some more areas.

Also, I want to know it will only work when a symbol has been saved from the backend office.

ashishkumar-webkul commented:
@devansh-webkul there are some conflicts with the PR, please check and update.

devansh-webkul commented:
This pull request has been here for too long. I am checking and updating this as per the requirement.

devansh-webkul commented:
@ashishkumar-webkul,

This pull request is ready for test again.

devansh-webkul commented:
This pull request will merge in the 2.1 branch. So switched from the master.

ashishkumar-webkul commented:
@devansh-webkul Sir

As checked and Found that there are some issues which are appeairng. Please check  the below details.
I have added migration as well.

1. I was able to add any values in the Decimal Seperator. Group Seperator and Decimal. 

**Video**

https://github.com/bagisto/bagisto/assets/92721837/a590ef06-d87a-4e84-a9f8-f5c9867f003b

2. I have added 4 decimal points and it is reflecting only 2 values.

**Video**

https://github.com/bagisto/bagisto/assets/92721837/d3a529d6-e578-4635-ab66-21a1502154a4

3. Proper Validation need to be added in Saving inappropriate values.

Screenshot -> **https://scrsht.com/d1fbd47c**

ashishkumar-webkul reviewed (commented):
Need to Make some changes related to  the issues.

Please follow the comment 

https://github.com/bagisto/bagisto/pull/8593#issuecomment-1994383236

devansh-webkul commented:
@ashishkumar-webkul,

- For the first point, possible decimal and group separators can vary depending on the specific language or context. Examples include comma (,), period (.), space, apostrophe ('), middle dot (·), hyphen (-), underscore (_), thin space, and non-breaking space. Therefore, it's better to keep the validation open so that users can add separators as per their needs. This isn't an issue because the admin person can configure it according to their requirements.

devansh-webkul commented:
@ashishkumar-webkul,

- The second point must be a cache issue. I verified it locally and it's working fine.

ashishkumar-webkul reviewed (approved):
Checked and Understood all the concerns related to the Demical and Seperator working and Decimal Value is appearing due to cache and hence I have varified the same and working fine. Hence I am approving the PR.
